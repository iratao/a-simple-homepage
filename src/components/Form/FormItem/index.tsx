import React, { useEffect, useState } from 'react'
import ErrorMsg from '../ErrorMsg'
import { FormInstance, Rules, FormItemType } from '../types'
import styles from './index.css'

interface FormItemProps {
  children: React.ReactElement;
  form?: FormInstance;
  name?: string;
  rules?: Rules;
  type?: FormItemType;
  error?: string;
  className?: string;
}

const FormItem = ({ children, name, form, rules, type, error, className }: FormItemProps) => {
  const [field, setField] = useState({})

  useEffect(() => {
    if (name && form) {
      setField({
        ...form.registerField({ name, rules, type })
      })
    }
  }, [])

  return (
    <div className={`${styles.formItem} ${className}`}>
      {React.Children.map(children, (child, index) => {
        if (child && index === 0 && form) {
          if (type === 'submit') {
            const onClick = child.props.onClick ? () => {
              child.props.onClick(form.validate())
            } : null
            return React.cloneElement(child, {
              ...field,
              ...child.props,
              onClick,
            })
          }
          if (name) {
            return React.cloneElement(child, {
              ...field,
              ...child.props,
            })
          }
        }
        return child;
      })}
      <ErrorMsg error={error} />
    </div>
  )
}

export default FormItem;