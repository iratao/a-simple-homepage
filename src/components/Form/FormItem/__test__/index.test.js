import React from 'react';
import renderer from 'react-test-renderer';
import FormItem from '../index'

it('renders correctly', () => {
  const tree = renderer
    .create(<FormItem><input /> </FormItem>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('register fields to it\'s child correctly', () => {
  const tree = renderer
    .create(<FormItem name="name"><input /> </FormItem>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

