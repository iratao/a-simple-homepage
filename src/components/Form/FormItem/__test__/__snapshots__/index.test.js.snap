// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`register fields to it's child correctly 1`] = `
<div
  className="formItem undefined"
>
  <input />
   
</div>
`;

exports[`renders correctly 1`] = `
<div
  className="formItem undefined"
>
  <input />
   
</div>
`;
